### PR TITLE
ENH: add support for reading binary Excel files (#15914)

### LINKED ISSUE #15914: read_excel does not work on excel file binary text or buffered binary text object
Python 3.5, Pandas 0.19.2

I have the following excel file, and I am trying to read its binary content, and then have read_excel read the binary content in, except this is not working, and there may be a bug somewhere, as I have specified the engine for reading.

>>> import pandas as pd
>>> f = open("Test_Prom_Data.xlsx", "rb")
>>> df = pd.read_excel(f.read(), engine = "xlrd")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/alexander/anaconda3/lib/python3.5/site-packages/pandas/io/excel.py", line 191, in read_excel
    io = ExcelFile(io, engine=engine)
  File "/home/alexander/anaconda3/lib/python3.5/site-packages/pandas/io/excel.py", line 251, in __init__
    raise ValueError('Must explicitly set engine if not passing in'
ValueError: Must explicitly set engine if not passing in buffer or path for io.
[Test_Prom_Data.xlsx](https://github.com/pandas-dev/pandas/files/901685/Test_Prom_Data.xlsx)

### PR DESCRIPTION
**Pull request by ftruzzi**
Hi, this is a quick fix for #15914.

After reading the issue I'm not sure if this is meant to be supported though, please let me know!

I ran into this issue when trying to pass to `read_excel` a binary file downloaded with `requests` so I think it would make sense to support this natively without the user having to figure out the need to call `io.BytesIO` first. If so, do we need to update other docstrings?

Thank you!

- [x] closes #15914
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Review by gfyoung — CHANGES_REQUESTED** [review]
This enhancement seems reasonable to me.

> If so, do we need to update other docstrings?

Yes!  Please update the docstring for `read_excel` as well as `io.rst` if necessary

cc @WillAyd 

Also, welcome to `pandas` ! 🚀

**Comment by ftruzzi** [comment]
Thank you @gfyoung, I have updated the `read_excel` docstring

File: doc/source/whatsnew/v1.0.0.rst [hunk 1]
```diff
@@ -816,6 +816,7 @@ I/O
 - Bug in :func:`read_json` where default encoding was not set to ``utf-8`` (:issue:`29565`)
 - Bug in :class:`PythonParser` where str and bytes were being mixed when dealing with the decimal field (:issue:`29650`)
 - :meth:`read_gbq` now accepts ``progress_bar_type`` to display progress bar while the data downloads. (:issue:`29857`)
+- :func:`read_excel` now accepts binary data (:issue:`15914`)
```
Comment (ftruzzi) [inline doc/source/whatsnew/v1.0.0.rst]: can anyone clarify the most appropriate between `:func:` and `:meth:`?

Comment (simonjayhawkins) [reply doc/source/whatsnew/v1.0.0.rst]: `:func:` is fine.

File: pandas/tests/io/excel/test_readers.py [hunk 2]
```diff
@@ -988,3 +988,14 @@ def test_conflicting_excel_engines(self, read_ext):
...
             with pytest.raises(ValueError, match=msg):
                 pd.read_excel(xl, engine="foo")
+
+    def test_excel_read_binary(self, engine, read_ext):
+        # GH 15914
+        expected = pd.read_excel("test1" + read_ext, engine=engine)
+
+        with open("test1" + read_ext, "rb") as f:
+            data = f.read()
+            f.close()
```
Comment (simonjayhawkins) [inline pandas/tests/io/excel/test_readers.py]: why is this necessary?

Comment (ftruzzi) [reply pandas/tests/io/excel/test_readers.py]: it's not, thank you as I didn't remember this

File: pandas/io/excel/_base.py [hunk 3]
```diff
@@ -350,6 +350,8 @@ def __init__(self, filepath_or_buffer):
             self.book = self.load_workbook(filepath_or_buffer)
         elif isinstance(filepath_or_buffer, str):
             self.book = self.load_workbook(filepath_or_buffer)
+        elif isinstance(filepath_or_buffer, bytes):
+            self.book = self.load_workbook(BytesIO(filepath_or_buffer))
```
Comment (WillAyd) [inline pandas/io/excel/_base.py]: Is BytesIO absolutely required here? Seems like we could treat str / bytes more consistently

**Review by WillAyd — CHANGES_REQUESTED** [review]
see question above. otherwise lgtm

File: pandas/io/excel/_base.py (hunk 3 above)
Comment (gfyoung) [reply pandas/io/excel/_base.py]: > Seems like we could treat str / bytes more consistently

@WillAyd : Consistently how?

Comment (WillAyd) [reply pandas/io/excel/_base.py]: We aren't wrapping str with an IO object so asking if that's really required for bytes

File: pandas/io/excel/_base.py [hunk 4]
```diff
@@ -40,7 +40,7 @@
 
 Parameters
 ----------
-io : str, ExcelFile, xlrd.Book, path object or file-like object
+io : str, ExcelFile, xlrd.Book, path object, binary data or file-like object
```
Comment (gfyoung) [inline pandas/io/excel/_base.py]:
```suggestion
io : str, ExcelFile, xlrd.Book, path object, binary data, or file-like object
```

Comment (gfyoung) [reply pandas/io/excel/_base.py]: BTW, I believe we still need an update to `io.rst`.

File: pandas/io/excel/_base.py (hunk 3 above)
Comment (gfyoung) [reply pandas/io/excel/_base.py]: Gotcha.  Thanks for clarifying!

@ftruzzi : Can you confirm if `BytesIO` is needed or not?

I suspect the answer is yes, but makes sense to double check for future reference.

Comment (ftruzzi) [reply pandas/io/excel/_base.py]: Yes I think it is needed. Probably `str` works just for filenames and not for stringified objects?

I had tried to treat `str` and `buffer` the same way, which could work with xlrd by passing the file contents to the `file_contents` kwarg in `xlrd.open_workbook` but I found no way to do the same thing in openpyxl and other engines.

And just by skipping `BytesIO`, engines complain as they try to open the thing as if it was a file:
```
File "C:\Users\ftruzzi\Documents\projects\bixby\capsule-dev-it\venv\lib\site-packages\xlrd\__init__.py", line 111, in open_workbook
    with open(filename, "rb") as f:
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x91 in position 14: invalid start byte
```
and
```
openpyxl.utils.exceptions.InvalidFileException: openpyxl does not support b'.relspk\x05\x06\x00\x00\x00\x00\x1d\x00\x1d\x00\xe1\x07\x00\x00\x8f\xa9\x0e\x00\x00\x00' file format, please check you can open it with Excel first. Supported formats are: .xlsx,.xlsm,.xltx,.xltm
```

Comment (ftruzzi) [reply pandas/io/excel/_base.py]: Yes I think it is needed. Probably `str` works just for filenames and not for stringified objects?

I had tried to treat `str` and `buffer` the same way, which could work with xlrd by passing the file contents to the `file_contents` kwarg in `xlrd.open_workbook` but I found no way to do the same thing in openpyxl and other engines.

And just by skipping `BytesIO`, engines complain as they try to open the thing as if it was a file:
```
File "...\xlrd\__init__.py", line 111, in open_workbook
    with open(filename, "rb") as f:
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x91 in position 14: invalid start byte
```
and
```
openpyxl.utils.exceptions.InvalidFileException: openpyxl does not support b'.relspk\x05\x06\x00\x00\x00\x00\x1d\x00\x1d\x00\xe1\x07\x00\x00\x8f\xa9\x0e\x00\x00\x00' file format, please check you can open it with Excel first. Supported formats are: .xlsx,.xlsm,.xltx,.xltm
```

File: pandas/io/excel/_base.py (hunk 4 above)
Comment (ftruzzi) [reply pandas/io/excel/_base.py]: what do you think should be added? an example of reading binary files?

Comment (gfyoung) [reply pandas/io/excel/_base.py]: On second look, I actually don't think we need anything here in `io.rst`.  Thanks for asking!

Also, looking at the documentation change here, now that I think about it, what do you think about just doing this:

```suggestion
io : str, bytes, ExcelFile, xlrd.Book, path object, or file-like object
```

**Review by WillAyd — APPROVED** [review]
Lgtm merge when happy @gfyoung

File: pandas/io/excel/_base.py (hunk 4 above)
Comment (ftruzzi) [reply pandas/io/excel/_base.py]: `bytes` looks much better but please let's keep the final comma out :P

Comment (gfyoung) [reply pandas/io/excel/_base.py]: > but please let's keep the final comma out :P

In this repository, we put commas in those situations.

Comment (ftruzzi) [reply pandas/io/excel/_base.py]: As you prefer and sorry if it sounded bad, didn't mean to! I had looked at other read_* docstrings so they might need updating as well.

**Comment by gfyoung** [comment]
Thanks @ftruzzi!

**Comment by ftruzzi** [comment]
Thank you @gfyoung @simonjayhawkins @WillAyd for your help and thorough reviews! Happy to contribute :)
